icicle_v3/backend/cpu/include/cpu_ntt.h

```diff
@@ -0,0 +1,331 @@
+
+  template <typename S>
+  class CpuNttDomain
+  {
+    int max_size = 0;
+    int max_log_size = 0;
+    std::unique_ptr<S[]> twiddles;
+    std::mutex domain_mutex;
+
+  public:
+    static eIcicleError
+    cpu_ntt_init_domain(const Device& device, const S& primitive_root, const NTTInitDomainConfig& config);
+
+    static eIcicleError cpu_ntt_release_domain(const Device& device);
+
+    template <typename U, typename E>
+    eIcicleError
+    cpu_ntt_ref(const Device& device, const E* input, int size, NTTDir dir, NTTConfig<S>& config, E* output);
+
+    template <typename U, typename E>
+    eIcicleError cpu_ntt(const Device& device, const E* input, int size, NTTDir dir, NTTConfig<S>& config, E* output);
+
+    const S* get_twiddles() const { return twiddles.get(); }
+    const int get_max_size() const { return max_size; }
+
+    static inline CpuNttDomain<S> s_ntt_domain;
+  };
+
+  template <typename S>
+  eIcicleError
+  CpuNttDomain<S>::cpu_ntt_init_domain(const Device& device, const S& primitive_root, const NTTInitDomainConfig& config)
+  {
+    // (1) check if need to refresh domain. TODO SHANIE - reusing domain for different ntt sizes if possible
+    if (s_ntt_domain.twiddles != nullptr) { return eIcicleError::SUCCESS; }
+
+    // Lock the mutex to ensure thread safety during initialization
+    std::lock_guard<std::mutex> lock(s_ntt_domain.domain_mutex); // Added for thread safety
+
+    // Double-checked locking
+    if (s_ntt_domain.twiddles == nullptr) {
+      // (2) build the domain
+
+      bool found_logn = false;
+      S omega = primitive_root;
+      const unsigned omegas_count = S::get_omegas_count();
+      for (int i = 0; i < omegas_count; i++) {
+        omega = S::sqr(omega);
+        if (!found_logn) {
+          ++s_ntt_domain.max_log_size;
+          found_logn = omega == S::one();
+          if (found_logn) break;
+        }
+      }
+
+      s_ntt_domain.max_size = (int)pow(2, s_ntt_domain.max_log_size);
+      if (omega != S::one()) {
+        ICICLE_LOG_ERROR << "Primitive root provided to the InitDomain function is not in the subgroup";
+        return eIcicleError::INVALID_ARGUMENT;
+      }
+
+      // calculate twiddles
+      // Note: radix-2 INTT needs ONE in last element (in addition to first element), therefore have n+1 elements
+
+      std::unique_ptr<S[]> temp_twiddles(new S[s_ntt_domain.max_size + 1]);
+
+      S tw_omega = primitive_root;
+      temp_twiddles[0] = S::one();
+      for (int i = 1; i <= s_ntt_domain.max_size; i++) {
+        temp_twiddles[i] = temp_twiddles[i - 1] * tw_omega;
+      }
+      s_ntt_domain.twiddles = std::move(temp_twiddles); // Assign twiddles using unique_ptr
+    }
+    return eIcicleError::SUCCESS;
+  }
+
+  template <typename S>
+  eIcicleError CpuNttDomain<S>::cpu_ntt_release_domain(const Device& device)
+  {
+    // No need to explicitly free memory since unique_ptr will handle it automatically
+    s_ntt_domain.twiddles.reset(); // Set twiddles to nullptr
```

Comment:
make sure to use the mutex here too. I think in CUDA it's missing but I think it is important

Reply:
fixed